rapid7/ruby_smb
# Fix for old samba compatibility

This fixes the issue described in https://github.com/rapid7/ruby_smb/issues/161 and other issues related to https://github.com/rapid7/metasploit-framework/issues/13976:
- Handle error packets with older Samba versions:
  It handles the case where an SMB2 error packet is expected, but the server sent an SMB1 empty packet instead. This behavior has been observed with older versions of Samba when something goes wrong on the server side.
- Handle pipe name backslash prefix according to the SMB version (see https://github.com/rapid7/ruby_smb/issues/161):
  Enforce backslash prefix for named pipe with SMB1 and remove backslash prefix for named pipe with SMB2 (if needed).
- Avoid double close_key in `winreg` enum keys and values methods.
- Improve RubySMB::Error::InvalidPacket logic.

## Verification
1. Install [Samba 3.6](https://gitlab.com/samba-team/samba/-/tree/v3-6-stable)
2. Do: `ruby examples/net_share_enum_all.rb <host> <username> <password> 1`
3. Verify it does not fail with a `STATUS_OBJECT_NAME_NOT_FOUND` error
4. Do: `ruby examples/enum_registry_key.rb <host> <username> <password> HKLM`
5. Verify it does not fail with `Not a Response packet (RubySMB::Dcerpc::Error::InvalidPacket)` error
